File: benchmarks/remediation/fixtures/js-command-module-function/tests/verify.js
// Two invocation modes.
//   node tests/verify.js --exploit | --behavior   Sandbox mode: checks the app.js in this tree.
//   node tests/verify.js <original> <repaired>    Harness mode: compares the checked-in trees.
const assert = require('node:assert/strict');
const Module = require('node:module');
const path = require('node:path');

const ATTACK = 'page.html; touch /tmp/owned';
const LEGITIMATE = 'page.html';

// Stands in for child_process so the test sees how the command was started without starting it.
// `shell` is the command line a shell would interpret: a string for exec, null for an argv call.
function withFakeChildProcess(modulePath, run) {
  const calls = [];
  const fake = {
    exec: (command, ...rest) => (calls.push({ shell: String(command), args: null }), rest.find((v) => typeof v === 'function')),
    execFile: (file, ...rest) => {
      const args = Array.isArray(rest[0]) ? rest[0] : [];
      calls.push({ shell: null, command: String(file), args: args.map(String) });
      return rest.find((v) => typeof v === 'function');
    },
  };
  fake.spawn = fake.execFile;
  const original = Module._load;
  Module._load = function (request, parent, isMain) {
    if (request === 'child_process' || request === 'node:child_process') return fake;
    return original.call(this, request, parent, isMain);
  };
  try {
    delete require.cache[require.resolve(modulePath)];
    run(require(modulePath));
  } finally {
    Module._load = original;
  }
  return calls[0];
}

const local = () => path.resolve(__dirname, '..', 'app.js');
const render = (target, name) => withFakeChildProcess(target, (m) => m.renderReport(name, () => {}));

function exploit() {
  const call = render(local(), ATTACK);
  if (typeof call.shell === 'string') {
    process.stderr.write('vulnerability present: the file name reaches a shell command string\n');
    process.exit(1);
  }
  assert.ok(call.args.includes(ATTACK), 'the payload must survive as its own argument');
  process.exit(0);
}

function behavior() {
  const call = render(local(), LEGITIMATE);
  const rendered = JSON.stringify(call);
  assert.ok(rendered.includes('wkhtmltopdf'));
  assert.ok(rendered.includes(LEGITIMATE));
  assert.ok(rendered.includes('report.pdf'));
  process.exit(0);
}

function compare(originalPath, repairedPath) {
  const original = render(originalPath, ATTACK);
  const repaired = render(repairedPath, ATTACK);
  const legitimate = render(repairedPath, LEGITIMATE);
  assert.match(original.shell, /touch \/tmp\/owned/);
  assert.equal(repaired.shell, null, 'the repair must not build a shell command string');
  assert.equal(repaired.command, 'wkhtmltopdf');
  assert.ok(repaired.args.includes(ATTACK));
  assert.ok(legitimate.args.includes(LEGITIMATE));
  process.stdout.write(JSON.stringify({ vulnerability_observed: true, behavior_preserved: true }));
}

const mode = process.argv[2];
if (mode === '--exploit') exploit();
else if (mode === '--behavior') behavior();
else compare(process.argv[2], process.argv[3]);

File: benchmarks/remediation/fixtures/python-command-os-system/tests/verify.py
"""Two invocation modes.

  python3 tests/verify.py --exploit | --behavior   Sandbox mode: checks the app.py in this tree.
  python3 tests/verify.py <original> <repaired>    Harness mode: compares the checked-in trees.
"""
import importlib.util
import json
import os
import subprocess
import sys

ATTACK = "/var/log/app.log; rm -rf /"
LEGITIMATE = "/var/log/app.log"
CALLS = []


def recording_system(command):
    CALLS.append({"args": command, "shell": True})
    return 0


def recording_run(args, **kwargs):
    CALLS.append({"args": args, "shell": bool(kwargs.get("shell"))})
    return None


os.system = recording_system
subprocess.run = recording_run


def load(path, name="app_under_test"):
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def local_path():
    return os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "app.py")


def exploit():
    del CALLS[:]
    load(local_path()).archive_log(ATTACK)
    call = CALLS[0]
    if call["shell"] or isinstance(call["args"], str):
        sys.stderr.write("vulnerability present: the argument reaches a shell command string\n")
        sys.exit(1)
    assert ATTACK in call["args"]
    sys.exit(0)


def behavior():
    del CALLS[:]
    assert load(local_path()).archive_log(LEGITIMATE) == LEGITIMATE + ".gz"
    rendered = json.dumps(CALLS[0]["args"])
    assert "gzip" in rendered and LEGITIMATE in rendered
    sys.exit(0)


def compare(original_path, repaired_path):
    del CALLS[:]
    load(original_path, "original").archive_log(ATTACK)
    assert CALLS[0]["shell"] is True
    assert "rm -rf /" in CALLS[0]["args"]
    del CALLS[:]
    load(repaired_path, "repaired").archive_log(ATTACK)
    assert CALLS[0] == {"args": ["gzip", "--force", ATTACK], "shell": False}
    del CALLS[:]
    load(repaired_path, "repaired_legit").archive_log(LEGITIMATE)
    assert CALLS[0]["args"] == ["gzip", "--force", LEGITIMATE]
    sys.stdout.write(json.dumps({"vulnerability_observed": True, "behavior_preserved": True}))


if __name__ == "__main__":
    mode = sys.argv[1] if len(sys.argv) > 1 else ""
    if mode == "--exploit":
        exploit()
    elif mode == "--behavior":
        behavior()
    else:
        compare(sys.argv[1], sys.argv[2])
